### DISCUSSION TITLE
Airflow Executor without MultiThreading

### DISCUSSION THREAD

**Discussion by micaelarg** [apache/airflow#27275] [Q&A]
I have a server in UpCloud, where I am running airflow. UpCloud does not support multithreading, thus billiard is not an option to parallelize (using python) in airflow. Any suggestion on how to parallelize without using billiard?

**Comment by potiuk** [comment]
I think if you need help, you should explain more precisely what you want to do and how you are doing It. No idea what 'lack of multithreading" is and why you "cannot use billiard" (especially that billiard is not multithreading but multiprocess and those are two completely different things). I do not even get what you  really want to parallelize and where. 

Airflow itself uses multiprocesing even with Local Executor, so I have no idea how it would work without mutlprocessing.

I am afraid  if you need help you need do better in explaining your problem statement and expectations.

**Reply by micaelarg** [reply]
I never wrote "lack of multithreading" in my question. Think I solved it, thanks.

**Reply by potiuk** [reply]
I wonder what was it. BTW. You wrote "UpCloud does not support multithreading"  :)

**Reply by micaelarg** [reply]
Yes, not the same. Regardless. It is exactly that (copied form upcloud support chat): "it looks like hyperthreading is disabled on our CPUs for security reasons. You may need to use multiprocessing rather than threading for running tasks in parallel.
e.g. https://jiafulow.github.io/blog/2020/12/30/running-python-tasks-in-parallel/" -- Unfortunately, since I added billiard, airflow doesn't even load. It does when I delete the dags with billiard, but it doesn't when I add them back. Any suggestion?

**Reply by micaelarg** [reply]
ps. It works just fine on my own computers + the server is newly set up

**Reply by potiuk** [reply]
No idea what it meas "airflow does not load". But maybe you can try to use constraints when you use ariflow  or external libraries - this is the only "supported" way of installing airflow. https://airflow.apache.org/docs/apache-airflow/stable/installation/installing-from-pypi.html

Also you shoudl not need to use "billiard" if you do not use CeleryWorkers - billiard is celery's replacement of mutliprocesing and it only shoudl be used if you use celery (because plain multiprocessing has issues if you use it with Celery).. 

Actually if you install Celery, then billiard should be already installed, so I think you either do not use Celery (and you can use plain multiprocessing from standard library) or you install billiard in some strange way.

**Reply by micaelarg** [reply]
Thank you - but how do you run parallel tasks/processes without billiard and localexcecutor then? A simple map(scraper, iterator)

**Reply by micaelarg** [reply]
everything I try i get the same error "AssertionError: daemonic processes are not allowed to have children"

**Reply by potiuk** [reply]
Then this is isue of UpCloud (if you are not using Celery). They are apparently putting you in that position. I think you want to eat cake and have it too. If they limit you with that,. they probably did not want you do it. And maybe they have some other safeguards in place (for example will not allow you to run multiple processes). If you ask me - change the deployment. And for sure it is borderline illegal for me to advise you how to bypass term&service limits of your provider (even if I would know how - but I don't).

**Reply by micaelarg** [reply]
thanks a lot.
